backend/LexBoxApi/Controllers/LoginController.cs

```diff
@@ -56,6 +58,67 @@ public class LoginController(
         return Redirect(returnTo);
     }
 
+    [HttpGet("google")]
+    [AllowAnonymous]
+    public IActionResult GoogleLogin(string? redirectTo = null)
+    {
+        if (string.IsNullOrEmpty(redirectTo)) redirectTo = "/home";
+        var authProps = new AuthenticationProperties { RedirectUri = redirectTo };
+        // If we want, we could look for expired-but-otherwise-valid JWT cookies and extract the user's email address, then:
+        // authProps.SetParameter("login_hint", email);
+        return Challenge(authProps, GoogleDefaults.AuthenticationScheme);
+    }
+
+    public async Task<string> CompleteGoogleLogin(ClaimsPrincipal? principal, string? returnTo)
+    {
+        returnTo ??= "/home";
+        var googleId = principal?.FindFirstValue(ClaimTypes.NameIdentifier);
+        var foundGoogleId = false;
+        var googleEmail = principal?.FindFirstValue(ClaimTypes.Email);
+        var googleName = principal?.FindFirstValue(ClaimTypes.Name);
+        var locale = principal?.FindFirstValue("locale");
+        var (authUser, userEntity) = await lexAuthService.GetUserByGoogleId(googleId);
+        if (authUser is not null)
+        {
+            foundGoogleId = true;
+        }
+        else
+        {
+            (authUser, userEntity) = await lexAuthService.GetUser(googleEmail);
+        }
+        if (authUser is null)
```

Comment:
One option would be to just delete/change the password if the email address isn't validated. But that would require an explanation.

Reply:
There's already code that checks that, from when I implemented the "Project invitation emails also register new users" feature. If the register page receives a JWT, then when it submits the form it also submits that JWT, so the backend receives a logged-in user token with audience of RegisterAccount. So when UserController.RegisterAccount runs, it checks if there's already a logged-in user. If there is, then we know that came from a valid JWT. And if there is a valid JWT, *and* the email on that JWT matches the email from the form, then we set the email address to already be validated. BUT if the email from the JWT doesn't match the one on the form, then the email isn't validated.

In other words, when we send someone to the register page and fill in their name and email address, we also create a token that says "We know the email address robin_munn@sil.org has been validated". If the user *changes* the email address on the form to be "rmunn@pobox.com" then we have to send out a validation email. But if the user submits the form without changing the email address, well, we got that email address from a JWT sent out by Google after the user logged in to Google. (Or we got it from the user clicking on a project-invite link that we sent out, and we know which email address we sent that link to). Either way, we know that they couldn't have gotten to the register page *with a JWT* without having already proven that either: 1) they can receive email at that account, or 2) Google believes they own that email address. Either one of those is good enough for us.

Or did you mean "Should we check whether Google's JWT includes `email_verified = true`, i.e. that Google has verified the email"? If that's what you meant, then https://jpassing.com/2021/01/27/what-does-the-email_verified-claim-indicate-in-google-idtokens/ suggests that the `email_verified` value in a Google JWT is pretty much always going to be true, so that step is a little meaningless.

Anyway, sorry to have written an essay. Short version: this specific token doesn't need to check if the email is validated, and the register page is already performing that check.